app/models/fe/question_sheet.rb

```diff
@@ -34,24 +34,20 @@ def self.new_with_page
 
     # count all questions including ones inside a grid
     def questions_count
-      parent_ids = elements.find_all{ |e| e.is_a?(Fe::QuestionGrid) }.collect(&:id)
-      count = questions.count
-      # grab all sub-elements in one query to speed things up a bit
-      while (sub_elements = Fe::Element.where(question_grid_id: parent_ids)).any?
-        count += sub_elements.to_a.count{ |e| e.is_a?(Fe::Question) }
-        parent_ids = sub_elements.find_all{ |e| e.is_a?(Fe::QuestionGrid) }.collect(&:id)
-      end
-      return count
-    end
-
-    def questions
-      pages.collect(&:questions).flatten
+      all_elements.questions.count
     end
 
+    # returns all elements including ones inside a grid
     def elements
       pages.collect(&:elements).flatten
     end
 
+    def all_elements
+      element_ids = pages.pluck(:all_element_ids).join(',').split(',')
+      element_ids << 0 # in case there are no elements, the query breaks at IN ()
+      Element.where(id: element_ids.split(','))
```

Comment:
You already split the element ids in line 46. I'm surprised this didn't throw an error calling split on an array. Maybe there isn't a test covering this?

Also, I'd lean towards the same method suggested above:
`element_ids.present? ? Element.where(id: element_ids) : Element.where("1 = 0")`